pkg/admissioncontroller/webhooks/auth/seed/graph/graph.go

```diff
@@ -0,0 +1,304 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graph
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	seedmanagementv1alpha1 "github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1"
+
+	"github.com/go-logr/logr"
+	gonumgraph "gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/simple"
+	"gonum.org/v1/gonum/graph/traverse"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Interface is used to track resources dependencies.
+type Interface interface {
+	// Setup registers the event handler functions for the various resource types.
+	Setup(ctx context.Context, c cache.Cache) error
+	// HasPathFrom returns true when there is a path from <from> to <to>.
+	HasPathFrom(fromType VertexType, fromNamespace, fromName string, toType VertexType, toNamespace, toName string) bool
+}
+
+type graph struct {
+	lock     sync.RWMutex
+	logger   logr.Logger
+	graph    *simple.DirectedGraph
+	vertices typeVertexMapping
+}
+
+var _ Interface = &graph{}
+
+// New creates a new graph interface for tracking resource dependencies.
+func New(logger logr.Logger) *graph {
+	return &graph{
+		logger:   logger,
+		graph:    simple.NewDirectedGraph(),
+		vertices: make(typeVertexMapping),
+	}
+}
+
+func (g *graph) Setup(ctx context.Context, c cache.Cache) error {
+	shootStates := &metav1.PartialObjectMetadata{}
+	shootStates.SetGroupVersionKind(gardencorev1alpha1.SchemeGroupVersion.WithKind("ShootState"))
```

Comment:
@timebertt I saw #3654, but I'm not using `PartialObjectMetadataList` but `PartialObjectMetadata` here. Do you think I need to change something?

Reply:
Nope, that would be the correct usage of `PartialObjectMetadata`, I guess.

Reply:
OK, thanks.